[ogr2ogr] Fix promote to multipart logic and add makevalid option

NyakudyaA opened the pull request:
Fixes https://github.com/qgis/QGIS/issues/58392

- [x] Add an option for makevalid, off by default
- [x] Raise an exception if Promote and  output geometry type are both selected
- [x] Raise an exception if Append and Overwrite are both selected

<img width="955" alt="Screenshot 2024-08-20 at 10 46 59" src="https://github.com/user-attachments/assets/df81f4ed-9498-451a-be23-29cefc80bece">

Commit:
Fix promote to multi logic and add makevalid option

agiudiceandrea commented:
Maybe it could be better to inform the user or raise an exception about the fact that they has selected the two incompatible options GTYPE and PROMOTETOMULTI (also when the algorithm is called by a Python script or it is used in a module).
It looks like the CONVERT_TO_LINEAR and CONVERT_TO_CURVE options are missing for the `-nlt` parameter, anyway they could be added as additional command line parameter.
Isn't the "Export to PostgreSQL (new connection)" (`gdal:importvectorintopostgisdatabasenewconnection`) also affected?
What about the "overwriting the existing table and appending to an existing table that should be mutually exclusive" reported in the same issue report?

Commit:
Fix review comments

NyakudyaA commented:
> Maybe it could be better to inform the user or raise an exception about the fact that they has selected the two incompatible options GTYPE and PROMOTETOMULTI (also when the algorithm is called by a Python script or it is used in a module). It looks like the CONVERT_TO_LINEAR and CONVERT_TO_CURVE options are missing for the `-nlt` parameter, anyway they could be added as additional command line parameter. Isn't the "Export to PostgreSQL (new connection)" (`gdal:importvectorintopostgisdatabasenewconnection`) also affected? What about the "overwriting the existing table and appending to an existing table that should be mutually exclusive" reported in the same issue report?

I updated the other script and added the missing import options for GTYPE. Now we get the following
<img width="955" alt="Screenshot 2024-08-20 at 13 37 13" src="https://github.com/user-attachments/assets/f7500f61-453d-4867-8fc0-44cef686c7d6">

agiudiceandrea commented:
Thanks!
According to https://gdal.org/programs/ogr2ogr.html#cmdoption-ogr2ogr-nlt it looks like that "-nlt CONVERT_TO_LINEAR and -nlt PROMOTE_TO_MULTI can be used simultaneously."

Commit:
update tests, fix append and overwrite options

agiudiceandrea commented:
@rouault it looks like the `-overwrite` and the `-append` options for ogr2ogr has been declared as mutually exclusive only since GDAL/OGR 3.9.0, rising the `Argument '-overwrite' not allowed with '-append'` error, while until GDAL/OGR 3.8.5 they could be both present at the same time without rising an error (in such case it seems to me only the last one in the command was taken in consideration). Is / was it the intended behaviour? Do you think the algorithm should allow `-append -overwrite` when used with GDAL/OGR < 3.9.0 and only rise an exception for GDAL/OGR >= 3.9.0, or should it always rise an exception regardless of the GDAL/OGR version used?

rouault commented:
-overwrite and -append have always been mutually exclusive. It is just that GDAL 3.9 starts making it an error. In previous versions, the last specified argument would win over the previously secified. It makes sense for QGIS to treat them as mutually exlusive for any GDAL version

NyakudyaA commented:
@agiudiceandrea, should we also handle the `-overwrite` and `-addfields` as mutually exclusive? Should we also add some validation logic for the `Additional creation options`, I assume if future gdal versions error in -overwrite and -append, then it should be handled or do we let users handle this since this option is assumed to be useful for advanced users

agiudiceandrea commented:
> should we also handle the `-overwrite` and `-addfields` as mutually exclusive?

It look like they are not actually enforced as mutually exclusive in ogr2ogr, but probably they should be.

> Should we also add some validation logic for the `Additional creation options`

AFAIK there is no validation in other algorithms for the additional creation options, but I've not actually checked.

Commit:
update tests, set ovewrite ti false

Commit:
update tests, remove extra space

Commit:
update tests, set promote to false

Commit:
update tests, use same input

Commit:
Format file after changes

NyakudyaA commented:
Not stale

NyakudyaA commented:
Not stale

NyakudyaA commented:
Not stale

nyalldawson commented on python/plugins/processing/algs/gdal/ogr2ogrtopostgislist.py:
```diff
@@ -316,7 +330,17 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
-        if promotetomulti:
+        if make_valid:
+            arguments.append('-makevalid')
+        if promotetomulti and len(self.GEOMTYPE[self.parameterAsEnum(parameters, self.GTYPE, context)]) > 0:
```
```suggestion
        if promotetomulti and self.GEOMTYPE[self.parameterAsEnum(parameters, self.GTYPE, context)]:
```

nyalldawson commented on python/plugins/processing/algs/gdal/ogr2ogrtopostgislist.py:
```diff
@@ -316,7 +330,17 @@ def getConsoleCommands(self, parameters, context, feedback, executing=True):
+            else:
+                raise QgsProcessingException(
+                    self.tr(
+                        'Only one of Promote to Multipart or Output Geometry Type (excluding Convert to Linear) can be enabled.'))
```
```suggestion
                        'Only one of "Promote to Multipart" or "Output geometry type" (excluding Convert to Linear) can be enabled.'))
```

nyalldawson committed:
Apply suggestions from code review

nyalldawson commented:
@agiudiceandrea was there more here you wanted to see?

Commit:
fix missing bracket reported by flake

agiudiceandrea commented:
Looks good to me.
